toolz/itertoolz.py

```diff
@@ -97,6 +98,32 @@ def groupby(key, seq):
     return rv
 
 
+def indices(*sizes):
+    """ Iterates over a length/shape.
+
+    >>> list(indices(3, 2))
+    [(0, 0), (0, 1), (1, 0), (1, 1), (2, 0), (2, 1)]
+
+    This can help nicely index an array.
+
+    >>> l = [[1, 2],
+    ...      [3, 4],
+    ...      [5, 6]]
+
+    >>> for i, j in indices(3, 2):
+    ...     print("l[%i][%i] = %i" % (i, j, l[i][j]))
+    l[0][0] = 1
+    l[0][1] = 2
+    l[1][0] = 3
+    l[1][1] = 4
+    l[2][0] = 5
+    l[2][1] = 6
```

Comment:
Sounds reasonable. I'm ok with not including it in the main namespace.

<hr>

Yeah `numpy.indices` is pretty different from this. Instead of doing something like this, it creates a massive array such that each index combination is specified. This ends up being pretty expensive for large arrays.

We can actually do much better if we note that much of this information is redundant and we are willing to part with having it in one big array. For most use cases, these are safe assumptions. Following them we get something like [this]( http://xnumpy.readthedocs.io/en/v0.1.0/xnumpy.core.html#xnumpy.core.indices ). For decent sized arrays, it is not unreasonable to see an order of magnitude or potentially a few orders of magnitude speed up by following this strategy.*

Even if we do need a full array with all combinations like `numpy.indices`, we can pack the result from the `xnumpy` function linked above into an array and still cutdown the creation time to roughly half.*

\* My benchmarking is still rather primitive at this point, but it does seem reliable thus far.